presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,575 @@
+            POWER_FIVES_INT[i] = BigInteger.valueOf(5).pow(i).intValue();
+        }
+    }
+
+    public static Slice unscaledDecimal()
+    {
+        return Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+    }
+
+    public static Slice unscaledDecimal(Slice decimal)
+    {
+        return Slices.copyOf(decimal);
+    }
+
+    public static Slice unscaledDecimal(String unscaledValue)
+    {
+        return unscaledDecimal(new BigInteger(unscaledValue));
+    }
+
+    public static Slice unscaledDecimal(BigInteger unscaledValue)
+    {
+        byte[] bytes = unscaledValue.abs().toByteArray();
+
+        if (bytes.length > UNSCALED_DECIMAL_128_SLICE_LENGTH
+                || (bytes.length == UNSCALED_DECIMAL_128_SLICE_LENGTH && (bytes[0] & SIGN_BYTE_MASK) != 0)) {
+            throwOverflowException();
+        }
+
+        // convert to little-endian order
+        reverse(bytes);
+
+        Slice decimal = Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        decimal.setBytes(0, bytes);
+        if (unscaledValue.signum() < 0) {
+            setNegative(decimal, true);
+        }
+
+        throwIfOverflows(decimal);
+
+        return decimal;
+    }
+
+    public static Slice unscaledDecimal(long unscaledValue)
+    {
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
+        boolean negative = isNegative(decimal);
+
+        if (hi != 0 || ((lo > Long.MIN_VALUE || !negative) && lo < 0)) {
+            throwOverflowException();
+        }
+
+        return negative ? -lo : lo;
+    }
+
+    public static void copyUnscaledDecimal(Slice from, Slice to)
+    {
+        setRawLong(to, 0, getRawLong(from, 0));
+        setRawLong(to, 1, getRawLong(from, 1));
+    }
+
+    public static Slice rescale(Slice decimal, int rescaleFactor)
+    {
+        if (rescaleFactor == 0) {
+            return decimal;
+        }
+        else if (rescaleFactor > 0) {
+            if (rescaleFactor >= POWERS_OF_TEN.length) {
+                throwOverflowException();
+            }
+            return multiply(decimal, POWERS_OF_TEN[rescaleFactor]);
+        }
+        else {
+            Slice result = unscaledDecimal();
+            scaleDownFive(decimal, -rescaleFactor, result);
```

Comment:
Add comment to source code to limit number of WTFs from future developer generations.

![](http://i2.wp.com/commadot.com/wp-content/uploads/2009/02/wtf.png?w=550)